File: test/supernode-storage.test.js
import test from 'node:test'
import assert from 'node:assert'
import crypto from 'crypto'
import fs from 'fs'
import path from 'path'
import os from 'os'
import { BlobStore } from '../reference-client/lib/blob-store.js'
import { SupernodeStorage } from '../supernode/storage/index.js'
import { generateEncryptionKey } from '../supernode/storage/mux/index.js'

function createTempDir () {
  return fs.mkdtempSync(path.join(os.tmpdir(), 'supernode-test-'))
}

function cleanup (dir) {
  fs.rmSync(dir, { recursive: true, force: true })
}

test('SupernodeStorage ingest and retrieve small file', async () => {
  const tempDir = createTempDir()
  try {
    const blobStore = new BlobStore(tempDir)
    const storage = new SupernodeStorage(blobStore)
    const masterKey = generateEncryptionKey()
    
    const originalData = Buffer.from('Hello, Supernode! This is a test file.')
    const fileName = 'test.txt'
    
    const { fileId, chunkHashes } = await storage.ingest(originalData, fileName, masterKey)
    
    assert.ok(fileId.startsWith('sha256:'), 'File ID should be sha256 hash')
    assert.strictEqual(chunkHashes.length, 1, 'Small file should have 1 chunk')
    
    const { data, fileName: recoveredName } = await storage.retrieve(fileId, masterKey)
    
    assert.deepStrictEqual(data, originalData, 'Retrieved data should match original')
    assert.strictEqual(recoveredName, fileName, 'File name should be preserved')
  } finally {
    cleanup(tempDir)
  }
})

test('SupernodeStorage ingest and retrieve large file (multi-chunk)', async () => {
  const tempDir = createTempDir()
  try {
    const blobStore = new BlobStore(tempDir)
    const storage = new SupernodeStorage(blobStore)
    const masterKey = generateEncryptionKey()
    
    const originalData = crypto.randomBytes(2.5 * 1024 * 1024)
    const fileName = 'large-file.bin'
    
    const { fileId, chunkHashes } = await storage.ingest(originalData, fileName, masterKey)
    
    assert.strictEqual(chunkHashes.length, 3, 'Should have 3 chunks for 2.5MB file')
    
    const { data } = await storage.retrieve(fileId, masterKey)
    
    assert.deepStrictEqual(data, originalData, 'Large file should roundtrip correctly')
  } finally {
    cleanup(tempDir)
  }
})

test('SupernodeStorage chunks are stored in blob store', async () => {
  const tempDir = createTempDir()
  try {
    const blobStore = new BlobStore(tempDir)
    const storage = new SupernodeStorage(blobStore)
    const masterKey = generateEncryptionKey()
    
    const originalData = Buffer.from('Test data for blob store verification')
    
    const { chunkHashes } = await storage.ingest(originalData, 'test.txt', masterKey)
    
    for (const hash of chunkHashes) {
      assert.ok(blobStore.has(hash), `Blob store should have chunk ${hash}`)
      const data = blobStore.get(hash)
      assert.ok(data, 'Chunk data should be retrievable')
      assert.ok(data.length > 0, 'Chunk should have content')
    }
  } finally {
    cleanup(tempDir)
  }
})

test('SupernodeStorage wrong master key fails retrieval', async () => {
  const tempDir = createTempDir()
  try {
    const blobStore = new BlobStore(tempDir)
    const storage = new SupernodeStorage(blobStore)
    const masterKey = generateEncryptionKey()
    const wrongKey = generateEncryptionKey()
    
    const originalData = Buffer.from('Secret data')
    const { fileId } = await storage.ingest(originalData, 'secret.txt', masterKey)
    
    await assert.rejects(async () => {
      await storage.retrieve(fileId, wrongKey)
    }, 'Wrong master key should fail retrieval')
  } finally {
    cleanup(tempDir)
  }
})

test('SupernodeStorage emits events during ingest', async () => {
  const tempDir = createTempDir()
  try {
    const blobStore = new BlobStore(tempDir)
    const storage = new SupernodeStorage(blobStore)
    const masterKey = generateEncryptionKey()
    
    const events = []
    storage.on('chunk-ingested', (e) => events.push({ type: 'chunk', ...e }))
    storage.on('file-ingested', (e) => events.push({ type: 'file', ...e }))
    
    const originalData = crypto.randomBytes(1.5 * 1024 * 1024)
    await storage.ingest(originalData, 'events-test.bin', masterKey)
    
    const chunkEvents = events.filter(e => e.type === 'chunk')
    const fileEvents = events.filter(e => e.type === 'file')
    
    assert.strictEqual(chunkEvents.length, 2, 'Should emit 2 chunk events')
    assert.strictEqual(fileEvents.length, 1, 'Should emit 1 file event')
    assert.strictEqual(fileEvents[0].chunks, 2, 'File event should report 2 chunks')
  } finally {
    cleanup(tempDir)
  }
})

test('SupernodeStorage stats includes manifest count', async () => {
  const tempDir = createTempDir()
  try {
    const blobStore = new BlobStore(tempDir)
    const storage = new SupernodeStorage(blobStore)
    const masterKey = generateEncryptionKey()
    
    await storage.ingest(Buffer.from('File 1'), 'file1.txt', masterKey)
    await storage.ingest(Buffer.from('File 2'), 'file2.txt', masterKey)
    
    const stats = storage.stats()
    
    assert.strictEqual(stats.manifestCount, 2, 'Should have 2 manifests')
    assert.strictEqual(stats.blobCount, 2, 'Should have 2 blobs')
  } finally {
    cleanup(tempDir)
  }
})

test('SupernodeStorage verifyChunkAsISO basic check', async () => {
  const tempDir = createTempDir()
  try {
    const blobStore = new BlobStore(tempDir)
    const storage = new SupernodeStorage(blobStore)
    const masterKey = generateEncryptionKey()
    
    const originalData = Buffer.from('Test data for ISO verification')
    const { chunkHashes } = await storage.ingest(originalData, 'iso-test.txt', masterKey)
    
    const isValid = storage.verifyChunkAsISO(chunkHashes[0])
    assert.ok(isValid, 'Chunk should pass basic ISO verification')
    
    const nonExistent = storage.verifyChunkAsISO('nonexistent-hash')
    assert.strictEqual(nonExistent, false, 'Non-existent chunk should fail')
  } finally {
    cleanup(tempDir)
  }
})

test('SupernodeStorage with erasure coding enabled', async () => {
  const tempDir = createTempDir()
  try {
    const blobStore = new BlobStore(tempDir)
    const storage = new SupernodeStorage(blobStore, {
      enableErasure: true,
      dataShards: 4,
      parityShards: 2
    })
    const masterKey = generateEncryptionKey()
    
    const originalData = Buffer.from('Erasure coding test data for redundancy verification')
    const fileName = 'erasure-test.txt'
    
    const { fileId, chunkHashes } = await storage.ingest(originalData, fileName, masterKey)
    
    assert.strictEqual(chunkHashes.length, 6, 'Should have 6 shards (4 data + 2 parity)')
    
    const { data, fileName: recoveredName } = await storage.retrieve(fileId, masterKey)
    
    assert.deepStrictEqual(data, originalData, 'Retrieved data should match original')
    assert.strictEqual(recoveredName, fileName, 'File name should be preserved')
  } finally {
    cleanup(tempDir)
  }
})

test('SupernodeStorage erasure coding recovers from missing shards', async () => {
  const tempDir = createTempDir()
  try {
    const blobStore = new BlobStore(tempDir)
    const storage = new SupernodeStorage(blobStore, {
      enableErasure: true,
      dataShards: 4,
      parityShards: 2
    })
    const masterKey = generateEncryptionKey()
    
    const originalData = Buffer.from('Recovery test data - this should survive shard loss')
    const { fileId, chunkHashes } = await storage.ingest(originalData, 'recovery.txt', masterKey)
    
    blobStore.delete(chunkHashes[0])
    blobStore.delete(chunkHashes[1])
    
    const { data } = await storage.retrieve(fileId, masterKey)
    
    assert.deepStrictEqual(data, originalData, 'Should recover data with 2 missing shards')
  } finally {
    cleanup(tempDir)
  }
})

test('SupernodeStorage erasure coding fails with too many missing shards', async () => {
  const tempDir = createTempDir()
  try {
    const blobStore = new BlobStore(tempDir)
    const storage = new SupernodeStorage(blobStore, {
      enableErasure: true,
      dataShards: 4,
      parityShards: 2
    })
    const masterKey = generateEncryptionKey()
    
    const originalData = Buffer.from('This will fail to recover')
    const { fileId, chunkHashes } = await storage.ingest(originalData, 'fail.txt', masterKey)
    
    blobStore.delete(chunkHashes[0])
    blobStore.delete(chunkHashes[1])
    blobStore.delete(chunkHashes[2])
    
    await assert.rejects(
      async () => storage.retrieve(fileId, masterKey),
      /Not enough shards/,
      'Should fail with 3 missing shards'
    )
  } finally {
    cleanup(tempDir)
  }
})

test('SupernodeStorage stats includes erasure config when enabled', async () => {
  const tempDir = createTempDir()
  try {
    const blobStore = new BlobStore(tempDir)
    const storage = new SupernodeStorage(blobStore, {
      enableErasure: true,
      dataShards: 8,
      parityShards: 4
    })
    
    const stats = storage.stats()
    
    assert.ok(stats.erasure, 'Stats should include erasure info')
    assert.strictEqual(stats.erasure.dataShards, 8)
    assert.strictEqual(stats.erasure.parityShards, 4)
    assert.strictEqual(stats.erasure.totalShards, 12)
  } finally {
    cleanup(tempDir)
  }
})

test('SupernodeStorage emits erasure events', async () => {
  const tempDir = createTempDir()
  try {
    const blobStore = new BlobStore(tempDir)
    const storage = new SupernodeStorage(blobStore, {
      enableErasure: true,
      dataShards: 4,
      parityShards: 2
    })
    const masterKey = generateEncryptionKey()
    
    const events = []
    storage.on('erasure-encoded', (e) => events.push({ type: 'encoded', ...e }))
    storage.on('erasure-decoded', (e) => events.push({ type: 'decoded', ...e }))
    
    const originalData = Buffer.from('Event test data')
    const { fileId } = await storage.ingest(originalData, 'events.txt', masterKey)
    
    assert.strictEqual(events.length, 1, 'Should emit erasure-encoded event')
    assert.strictEqual(events[0].type, 'encoded')
    assert.strictEqual(events[0].dataShards, 4)
    assert.strictEqual(events[0].parityShards, 2)
    
    await storage.retrieve(fileId, masterKey)
    
    assert.strictEqual(events.length, 2, 'Should emit erasure-decoded event')
    assert.strictEqual(events[1].type, 'decoded')
    assert.strictEqual(events[1].presentShards, 6)
  } finally {
    cleanup(tempDir)
  }
})

File: supernode-java/src/test/java/io/supernode/storage/SupernodeStorageTest.java
package io.supernode.storage;

import io.supernode.storage.isoforge.SizePreset;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import java.nio.charset.StandardCharsets;
import java.security.SecureRandom;
import java.util.Arrays;
import java.util.concurrent.atomic.AtomicInteger;

import static org.junit.jupiter.api.Assertions.*;

@DisplayName("SupernodeStorage")
class SupernodeStorageTest {
    
    private InMemoryBlobStore blobStore;
    private byte[] masterKey;
    
    @BeforeEach
    void setUp() {
        blobStore = new InMemoryBlobStore();
        masterKey = new byte[32];
        Arrays.fill(masterKey, (byte) 0x42);
    }
    
    @Nested
    @DisplayName("Basic Ingest/Retrieve")
    class BasicOperationsTests {
        
        @Test
        @DisplayName("should ingest and retrieve small file")
        void ingestAndRetrieveSmallFile() {
            SupernodeStorage storage = new SupernodeStorage(blobStore);
            
            byte[] content = "Hello, Supernode!".getBytes(StandardCharsets.UTF_8);
            String fileName = "hello.txt";
            
            SupernodeStorage.IngestResult ingestResult = storage.ingest(content, fileName, masterKey);
            
            assertNotNull(ingestResult.fileId());
            assertTrue(ingestResult.fileId().startsWith("sha256:"));
            assertFalse(ingestResult.chunkHashes().isEmpty());
            assertNotNull(ingestResult.encryptedManifest());
            
            SupernodeStorage.RetrieveResult retrieveResult = storage.retrieve(ingestResult.fileId(), masterKey);
            
            assertEquals(fileName, retrieveResult.fileName());
            assertEquals(content.length, retrieveResult.fileSize());
            assertArrayEquals(content, retrieveResult.data());
        }
        
        @Test
        @DisplayName("should ingest and retrieve large file (multi-chunk)")
        void ingestAndRetrieveLargeFile() {
            SupernodeStorage storage = new SupernodeStorage(blobStore);
            
            byte[] content = new byte[3 * 1024 * 1024];
            new SecureRandom().nextBytes(content);
            String fileName = "large-file.bin";
            
            SupernodeStorage.IngestResult ingestResult = storage.ingest(content, fileName, masterKey);
            
            assertTrue(ingestResult.chunkHashes().size() >= 3, "Should have at least 3 chunks for 3MB");
            
            SupernodeStorage.RetrieveResult retrieveResult = storage.retrieve(ingestResult.fileId(), masterKey);
            
            assertEquals(fileName, retrieveResult.fileName());
            assertArrayEquals(content, retrieveResult.data());
        }
        
        @Test
        @DisplayName("should handle empty file")
        void handleEmptyFile() {
            SupernodeStorage storage = new SupernodeStorage(blobStore);
            
            byte[] content = new byte[0];
            String fileName = "empty.txt";
            
            SupernodeStorage.IngestResult ingestResult = storage.ingest(content, fileName, masterKey);
            SupernodeStorage.RetrieveResult retrieveResult = storage.retrieve(ingestResult.fileId(), masterKey);
            
            assertEquals(0, retrieveResult.data().length);
        }
        
        @Test
        @DisplayName("should handle file exactly chunk-sized")
        void handleExactChunkSize() {
            SupernodeStorage storage = new SupernodeStorage(blobStore);
            
            byte[] content = new byte[SupernodeStorage.CHUNK_SIZE];
            new SecureRandom().nextBytes(content);
            
            SupernodeStorage.IngestResult ingestResult = storage.ingest(content, "exact.bin", masterKey);
            
            assertEquals(1, ingestResult.chunkHashes().size());
            
            SupernodeStorage.RetrieveResult retrieveResult = storage.retrieve(ingestResult.fileId(), masterKey);
            assertArrayEquals(content, retrieveResult.data());
        }
    }
    
    @Nested
    @DisplayName("Erasure Coding")
    class ErasureCodingTests {
        
        @Test
        @DisplayName("should ingest with erasure coding enabled")
        void ingestWithErasure() {
            SupernodeStorage.StorageOptions options = SupernodeStorage.StorageOptions.withErasure(4, 2);
            SupernodeStorage storage = new SupernodeStorage(blobStore, options);
            
            byte[] content = new byte[10000];
            new SecureRandom().nextBytes(content);
            
            SupernodeStorage.IngestResult ingestResult = storage.ingest(content, "erasure-test.bin", masterKey);
            
            assertTrue(ingestResult.chunkHashes().size() >= 6, "Should have shards (4 data + 2 parity)");
            
            SupernodeStorage.RetrieveResult retrieveResult = storage.retrieve(ingestResult.fileId(), masterKey);
            assertArrayEquals(content, retrieveResult.data());
        }
        
        @Test
        @DisplayName("stats should reflect erasure configuration")
        void statsReflectErasure() {
            SupernodeStorage.StorageOptions options = SupernodeStorage.StorageOptions.withErasure(4, 2);
            SupernodeStorage storage = new SupernodeStorage(blobStore, options);
            
            SupernodeStorage.StorageStats stats = storage.stats();
            
            assertNotNull(stats.erasure());
            assertEquals(4, stats.erasure().dataShards());
            assertEquals(2, stats.erasure().parityShards());
            assertEquals(6, stats.erasure().totalShards());
        }
    }
    
    @Nested
    @DisplayName("Events")
    class EventTests {
        
        @Test
        @DisplayName("should emit chunk ingested events")
        void emitChunkIngestedEvents() {
            SupernodeStorage storage = new SupernodeStorage(blobStore);
            
            AtomicInteger chunkCount = new AtomicInteger(0);
            storage.setOnChunkIngested(event -> {
                assertNotNull(event.fileId());
                assertNotNull(event.chunkHash());
                assertTrue(event.progress() >= 0 && event.progress() <= 1);
                chunkCount.incrementAndGet();
            });
            
            byte[] content = new byte[2 * SupernodeStorage.CHUNK_SIZE + 1000];
            new SecureRandom().nextBytes(content);
            
            storage.ingest(content, "chunked.bin", masterKey);
            
            assertEquals(3, chunkCount.get());
        }
        
        @Test
        @DisplayName("should emit file ingested event")
        void emitFileIngestedEvent() {
            SupernodeStorage storage = new SupernodeStorage(blobStore);
            
            AtomicInteger eventCount = new AtomicInteger(0);
            storage.setOnFileIngested(event -> {
                assertEquals("test.txt", event.fileName());
                assertTrue(event.size() > 0);
                assertTrue(event.chunkCount() > 0);
                eventCount.incrementAndGet();
            });
            
            storage.ingest("Test content".getBytes(), "test.txt", masterKey);
            
            assertEquals(1, eventCount.get());
        }
        
        @Test
        @DisplayName("should emit file retrieved event")
        void emitFileRetrievedEvent() {
            SupernodeStorage storage = new SupernodeStorage(blobStore);
            
            byte[] content = "Retrieve test".getBytes();
            SupernodeStorage.IngestResult ingestResult = storage.ingest(content, "retrieve.txt", masterKey);
            
            AtomicInteger eventCount = new AtomicInteger(0);
            storage.setOnFileRetrieved(event -> {
                assertEquals("retrieve.txt", event.fileName());
                assertEquals(content.length, event.size());
                eventCount.incrementAndGet();
            });
            
            storage.retrieve(ingestResult.fileId(), masterKey);
            
            assertEquals(1, eventCount.get());
        }
    }
    
    @Nested
    @DisplayName("Statistics")
    class StatisticsTests {
        
        @Test
        @DisplayName("should track blob count")
        void trackBlobCount() {
            SupernodeStorage storage = new SupernodeStorage(blobStore);
            
            SupernodeStorage.StorageStats initialStats = storage.stats();
            assertEquals(0, initialStats.blobCount());
            
            storage.ingest("File 1".getBytes(), "file1.txt", masterKey);
            storage.ingest("File 2".getBytes(), "file2.txt", masterKey);
            
            SupernodeStorage.StorageStats finalStats = storage.stats();
            assertEquals(2, finalStats.blobCount());
        }
        
        @Test
        @DisplayName("should track manifest count")
        void trackManifestCount() {
            SupernodeStorage storage = new SupernodeStorage(blobStore);
            
            SupernodeStorage.StorageStats initialStats = storage.stats();
            assertEquals(0, initialStats.manifestCount());
            
            storage.ingest("File 1".getBytes(), "file1.txt", masterKey);
            storage.ingest("File 2".getBytes(), "file2.txt", masterKey);
            
            SupernodeStorage.StorageStats finalStats = storage.stats();
            assertEquals(2, finalStats.manifestCount());
        }
        
        @Test
        @DisplayName("should report ISO size preset")
        void reportISOSizePreset() {
            SupernodeStorage storage = new SupernodeStorage(blobStore);
            
            SupernodeStorage.StorageStats stats = storage.stats();
            assertEquals("nano", stats.isoSize());
        }
    }
    
    @Nested
    @DisplayName("Manifest Operations")
    class ManifestOperationsTests {
        
        @Test
        @DisplayName("should store and retrieve manifest")
        void storeAndRetrieveManifest() {
            SupernodeStorage storage = new SupernodeStorage(blobStore);
            
            SupernodeStorage.IngestResult ingestResult = storage.ingest("Test".getBytes(), "test.txt", masterKey);
            
            assertTrue(storage.getManifest(ingestResult.fileId()).isPresent());
            assertArrayEquals(ingestResult.encryptedManifest(), 
                storage.getManifest(ingestResult.fileId()).get());
        }
        
        @Test
        @DisplayName("should return empty for unknown manifest")
        void emptyForUnknownManifest() {
            SupernodeStorage storage = new SupernodeStorage(blobStore);
            
            assertTrue(storage.getManifest("sha256:unknown").isEmpty());
        }
        
        @Test
        @DisplayName("should allow external manifest storage")
        void allowExternalManifestStorage() {
            SupernodeStorage storage = new SupernodeStorage(blobStore);
            
            byte[] fakeManifest = "fake-manifest-data".getBytes();
            storage.storeManifest("external-file-id", fakeManifest);
            
            assertTrue(storage.getManifest("external-file-id").isPresent());
            assertArrayEquals(fakeManifest, storage.getManifest("external-file-id").get());
        }
    }
    
    @Nested
    @DisplayName("Error Handling")
    class ErrorHandlingTests {
        
        @Test
        @DisplayName("should fail retrieve with wrong master key")
        void failRetrieveWithWrongKey() {
            SupernodeStorage storage = new SupernodeStorage(blobStore);
            
            byte[] content = "Secret data".getBytes();
            SupernodeStorage.IngestResult ingestResult = storage.ingest(content, "secret.txt", masterKey);
            
            byte[] wrongKey = new byte[32];
            Arrays.fill(wrongKey, (byte) 0xFF);
            
            assertThrows(Exception.class, () -> {
                storage.retrieve(ingestResult.fileId(), wrongKey);
            });
        }
        
        @Test
        @DisplayName("should fail retrieve for unknown file ID")
        void failRetrieveUnknownFileId() {
            SupernodeStorage storage = new SupernodeStorage(blobStore);
            
            assertThrows(IllegalArgumentException.class, () -> {
                storage.retrieve("sha256:unknown", masterKey);
            });
        }
    }
    
    @Nested
    @DisplayName("Determinism")
    class DeterminismTests {
        
        @Test
        @DisplayName("same content should produce same file ID")
        void sameContentSameFileId() {
            SupernodeStorage storage = new SupernodeStorage(blobStore);
            
            byte[] content = "Deterministic content".getBytes();
            
            SupernodeStorage.IngestResult result1 = storage.ingest(content.clone(), "file1.txt", masterKey);
            
            blobStore.clear();
            
            SupernodeStorage.IngestResult result2 = storage.ingest(content.clone(), "file2.txt", masterKey);
            
            assertEquals(result1.fileId(), result2.fileId(), "Same content should produce same file ID");
        }
    }
}
